pkg/datapath/loader/loader.go

```diff
@@ -298,34 +292,45 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 			// this log message should only represent failures with respect to
 			// loading the program.
 			if ctx.Err() == nil {
-				scopedLog.WithError(err).Warn("JoinEP: Failed to attach ingress program")
+				scopedLog.WithError(err).Warnf("JoinEP: Failed to attach %s program", progDirection)
 			}
+		}
+		finalizeChan <- finalize
+		return err
+	}
+}
+
+func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs *directoryInfo) error {
+	// Replace the current program
+	objPath := path.Join(dirs.Output, endpointObj)
+
+	if ep.IsHost() {
+		objPath = path.Join(dirs.Output, hostEndpointObj)
+		if err := l.reloadHostDatapath(ctx, ep, objPath); err != nil {
 			return err
 		}
-		defer finalize()
+	} else {
+		finalizers := make(chan func(), 2)
+		wg := &errgroup.Group{}
+		wg.Go(replaceDatapathRunner(ctx, ep, objPath, symbolFromEndpoint, dirIngress, finalizers))
 
 		if ep.RequireEgressProg() {
-			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, "")
-			if err != nil {
-				scopedLog := ep.Logger(Subsystem).WithFields(logrus.Fields{
-					logfields.Path: objPath,
-					logfields.Veth: ep.InterfaceName(),
-				})
-				// Don't log an error here if the context was canceled or timed out;
-				// this log message should only represent failures with respect to
-				// loading the program.
-				if ctx.Err() == nil {
-					scopedLog.WithError(err).Warn("JoinEP: Failed to attach egress program")
-				}
-				return err
-			}
-			defer finalize()
+			wg.Go(replaceDatapathRunner(ctx, ep, objPath, symbolToEndpoint, dirEgress, finalizers))
 		} else {
 			err := RemoveTCFilters(ep.InterfaceName(), netlink.HANDLE_MIN_EGRESS)
 			if err != nil {
 				log.WithField("device", ep.InterfaceName()).Error(err)
 			}
 		}
+
+		if err := wg.Wait(); err != nil {
+			return err
+		}
+
+		close(finalizers)
+		for finalize := range finalizers {
+			defer finalize()
+		}
```

Comment:
In the original code if one program replaces successfully it's finalizer will be called and the function will proceed to the second one.
After the change both of them are process in parallel.

Should we still defer finalizer for replacements that finished successfully or only if all replacement succeeded?

Reply:
It seems fine to me, but I can easily change if we need to finalize successful replacement. @ti-mo could you comment on this one, please?

Reply:
@alan-kut Good catch.

> if one program replaces successfully it's finalizer will be called

If the first prog attaches successfully, it will be deferred, not immediately called.

> Should we still defer finalizer for replacements that finished successfully or only if all replacement succeeded?

Yes, the first replacement might return a finalizer and the second one could fail. The finalizer of the first one still needs to be run. The correct ordering here should be:

```
err := wg.Wait()
<close & range finalizers>
if err != nil {
  return err
}
```